enhancements/monitoring/alerting-consistency.md

```diff
@@ -0,0 +1,195 @@
+be faced with a disaster.
+
+An example of something that is NOT a critical alert:
+* MCO and/or related components are completely dead/crash looping.
+
+Even though the above is quite significant, there is no risk of immediate loss
+of the cluster.
+
+You might be thinking to yourself: "But my component is **super** important, and
+if it's not working, the user can't do X, which user definitely will want."  All
+that is probably true, but that doesn't make your alert "critical" it just makes
+it worthy of an alert.
+
+### Warning Alerts
+
+TL/DR: fix this soon, or some things won't work and upgrades will probably be
+blocked.
+
+If your alert does not meet the criteria in "Critical Alerts" above, it belongs
+to the warning level or lower.
+
+Most alerts probably fit in this category.  This level should represent degraded
+or unavailable services that do not jeopardize the immediate life of cluster.
+
+Timeline:  ~60 minutes
+
+60 minutes?  That seems high!  That's because it is high.  We want to reduce
+the noise.  We've done a lot to make clusters and operators auto-heal
+themselves.  The whole idea is that if a condition has persisted for more than
+60 minutes, it's unlikely to be resolved without intervention any time later.
+
+#### Example 1: All machine-config-daemons are crashlooping.
+
+That seems significant
+and it probably is, but nobody has to get out of bed to fix this **right now**.
+But, what if a machine dies, I won't be able to get a replacement?  Yeah, that
+is probably true, but how often does that happen?  Also, that's an entirely
+separate set of concerns.
+
+#### Example 2:  I have an operator that needs at least 3 of something to work.
+* With 2/3 replicas, warning alert after 60M.
+* With 1/3 replicas, warning alert after 10M.
+* With 0/3 replicas, warning alert after 5M.
+
+Q: How long should we wait until the above conditions rises to a 'critical'
+alert?
+
+A: It should never rise to the level of a critical alert, it's not critical.  If
+it was, this section would not apply.
+
+#### Example 3:  I have an operator that only needs 1 replica to function
+
+Well, if the cluster can upgrade with only 1 replica, and the the service is
+not unavailable despite other replicas being unavailable, this can probably
+be just an info-level alert.  We should have a "minor" alert, TBD.  But for now,
+"info" is the closest thing.
+
+#### Q: What if a particular condition will block upgrades?
+
+A: It's a warning level alert.
+
+
+### Documentation Required
+
+Every alert needs to have a URL to relevant documentation to resolve the issue.
+
+If there's not documentation, there shouldn't be an alert.
+
+Sometimes an alert's impact is not obvious.  We should state the impact of the
+alert.  Don't expect users to understand the context, we have too many moving
+pieces.
+
+
+### Open Questions [optional]
+
+Frankly, having just "Critical" and "Warning" is not helpful.
```

Comment:
Can you give some reasons here why not? :)

Reply:
Inevitably, there are going to be some 'warning' level alerts that people won't care about.  We need a lower level that indicates that there is a problem, but the impact is low.  This would improve the signal to noise ratio of warning level alerts.

Reply:
Warning level already should create a ticket, so this is already low level enough, it should not page anyone so there is no noise really in an ideal scenario. If folks don't care about those warning alerts they can silence them after getting to a ticket.

Reply:
The problem is, we want all warning level alerts to be cared about and fixed in most situations.  This is about signal to noise.  If people are trained to ignore warning level alerts by default, they will miss important alerts.  This is why we need a 'minor' alert.

Alternatively, we can just not alert for 'minor' alerts.

Reply:
IMO with minor severity alerts, we're not avoiding the issue of folks not paying attention to alerts because they're noisy... If a warning alert is too noisy, it is a bug: the alert should either be removed or modified to be more accurate.
The Prometheus community has discussed for a long time about this: the (relative) consensus is that there should be a level for "everything's on fire" alerts and "please have a look at this" alerts. Informative alerts can be useful sometimes but should probably be the exception.

Reply:
For me, it's about prioritization.  Every alert should be investigated.  But, if you have 10 alerts all of the same level, how do you triage? EG: all of MCO is down, and you have one worker instance stopped in the cloud.  Today, both of these are alerts.  Is one of these situations more urgent than the others?  Do either of these require you to get out of bed in the middle of the night?

Reply:
Again if an alert is "critical" then yes you need to get out of the bed. Anything else can wait until somebody wakes up.
I'm not convinced that having "warning" and "less warning" alerts would work well given that we're already having a hard time classifying between warning and critical. Obviously I can be wrong and I'd listen carefully how folks on call feel about this...

Reply:
I have thought about this overnight, and for me it comes back to triage.  The group of critical alerts should be small, very well defined, highly documented and polished.

The group of warning alerts are the next priority.  If you find yourself managing many clusters with many alerts, you need a sensible way to triage these.  Clearly an API replica being down is more important than a daemonset not running on an individual worker.  They're both problems, but one should probably be investigated before the other.

I think, another way to state it is, think of a single alert in the context of the cluster, rather than the context of a single operator.  We all know there are 'minor' alerts, we just don't label them as minor.

Reply:
"Everything else can wait" yes, that's true.  But if you're managing dozens of clusters, you need a way to triage alerts, and you probably want to send a lot of stuff to /dev/null.  Our operation teams probably already have rules about which rules to care about and which rules to ignore.  Some of this particularly revolves around upgrades and asking "Should I fix this prior to upgrade?"  I want to remove that question.

Reply:
Even if we introduce a "minor" level, there would be people complaining that alert X should "warning" while it's "minor" (and vice-versa). At some point you will leverage additional alert labels to group/route the alerts to the correct destination (this is what OSD is doing from what I understand).
And triaging doesn't necessarily happen at the cluster level: if you run significant infrastructure, you probably have a PagerDuty-like service for dealing with on-call activities.

Reply:
I'm not talking about triaging on-call events (eg, critical).  Those are the ones that get paged and hopefully resolved.  I'm talking about all the stuff that fits into 'warning.'

I agree, some people will want alerts to change from one severity to another.  That's the entire point of this document, to establish **clear** guidelines.  Definitely generic alerts like Kube* I mentioned elsewhere, if we don't disable those entirely (we probably should), every one of those should be 'minor.'  It's a problem, but it's likely caused by a warning or higher level alert (or should be).  If something is down, and the service is not impacted, and losing more of them doesn't jeopardize the cluster's health, then it's probably okay to be minor.

Reply:
I would like this sentence added:

"The group of critical alerts should be small, very well defined, highly documented, polished and with a high bar set for entry.  This includes a mandatory review of a proposed critical alert by the Red Hat SRE team."

Having this would be an important checkpoint for SRE, one we have a workflow already in place to support you.